<a href="https://colab.research.google.com/github/ElizabethGarrison/Working-With-Data-Fundamentals/blob/main/MergingData.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Merging Data Sources

In this notebook, I've demonstrated one way to merge data sources using pandas DataFrames. In this case, I have three global social indicators datasets, all of which contain the same country names, which I can use as an index to zip (or glue) the files together. I'm using the **concat()** function with the **axis=1** flag in order to add variables rather than cases.

I use the **na_values[np.nan]** flag to deal with missing values in the Excel files. When I create the analytical model, I use the **missing='drop'** flag to exclude those cases from the analysis.

In [ ]:
# Code block 1a: Libraries
import pandas as pd
import numpy as np
import statsmodels.api as sm
import seaborn as sb
import matplotlib.pyplot as plt


In [ ]:
# Code block 2a: Reading an Excel spreadsheet to combine
GlobalIndicators1 = pd.read_excel('http://data.shortell.nyc/files/HumanDevelopment.xlsx', index_col='Country', na_values=[np.nan])
GlobalIndicators1.head()

HDI Rank  ...  GNI per Capita Rank Minus HDI Rank
Country                ...                                    
Norway              1  ...                                   5
Australia           2  ...                                  17
Switzerland         3  ...                                   6
Denmark             4  ...                                  11
Netherlands         5  ...                                   9

[5 rows x 7 columns]

In [ ]:
# Code block 2b: Reading another Excel spreadsheet
GlobalIndicators2 = pd.read_excel('http://data.shortell.nyc/files/GenderDevelopment.xlsx', index_col='Country', na_values=['NA'])
GlobalIndicators2.head()

GDI Rank  ...  Estimated Gross National Income per Capita (Male)
Country                ...                                                   
Norway              1  ...                                            72825.0
Australia           2  ...                                            50914.0
Switzerland         3  ...                                            69077.0
Denmark             4  ...                                            51727.0
Netherlands         5  ...                                            61641.0

[5 rows x 12 columns]

In [ ]:
# Code block 2c: Reading a third Excel spreadsheet
GlobalIndicators3 = pd.read_excel('http://data.shortell.nyc/files/GenderInequality.xlsx', index_col='Country', na_values=['NA'])
GlobalIndicators3.head()

GII Rank  ...  Labour Force Participation Rate (Male)
Country                ...                                        
Norway              1  ...                                    68.7
Australia           2  ...                                    71.8
Switzerland         3  ...                                    74.9
Denmark             4  ...                                    66.4
Netherlands         5  ...                                    70.6

[5 rows x 9 columns]

In [ ]:
# Code block 2d: Reading a fourth spreadsheet
GlobalIndicators4 = pd.read_csv('http://data.shortell.nyc/files/world-happiness-report-2019.csv', index_col='Country (region)')
GlobalIndicators4.head()

Ladder  ...  Healthy life\nexpectancy
Country (region)          ...                          
Finland                1  ...                      27.0
Denmark                2  ...                      23.0
Norway                 3  ...                      12.0
Iceland                4  ...                      13.0
Netherlands            5  ...                      18.0

[5 rows x 10 columns]

Here, we will combine the three new DataFrames we've just created into a new DataFrame by concatinating the parts. The flag, **axis=1**, allows us to stack the three files side-by-side, as it were, rather than on top of each other. We can do this because all of them are indexed by Country. We can now use variables from any of the data sources in our analysis.

In [ ]:
# Code block 3: Merging the DataFrames using the concat method in pandas
GlobalIndicatorsTotal = pd.concat([GlobalIndicators1, GlobalIndicators2, GlobalIndicators3, GlobalIndicators4], axis=1)
GlobalIndicatorsTotal.head()

HDI Rank  ...  Healthy life\nexpectancy
Norway            1.0  ...                      12.0
Australia         2.0  ...                      10.0
Switzerland       3.0  ...                       4.0
Denmark           4.0  ...                      23.0
Netherlands       5.0  ...                      18.0

[5 rows x 38 columns]

In our model, we are trying to explain variation in aggregate quality of life. We're going to use five predictors, all measuring social aspects of gender inequality. That is, social policy contributes to why some nations are higher on these dimensions than others. In the linear model, we are using each of these predictors as control variables for the others. In this way, we get an estimate of the unique effect of each predictor.

This is important, because these proxies of social policy intersect. That is, nations that have policy to encourage and support education for girls and women tend to have policies that give women more control over their reproductive rights, encourage labor force participation, and have space in the political sphere for women to run for elected office. But in the model, we can estimate the unique effects--the variation that a particular predictor can account for that the others cannot.

In [ ]:
# Code block 4: Using the merged data in a linear model
Y = GlobalIndicatorsTotal['Ladder'] # A measure of overall quality of life
X = GlobalIndicatorsTotal[['Mean Years of Education (Female)', 'Labour Force Participation Rate (Female)', 'Adolescent Birth Rate', 'Percent Representation in Parliament']]
X = sm.add_constant(X)
model0 = sm.OLS(Y, X, missing='drop').fit()
print(model0.summary())

                            OLS Regression Results                            
Dep. Variable:                 Ladder   R-squared:                       0.541
Model:                            OLS   Adj. R-squared:                  0.527
Method:                 Least Squares   F-statistic:                     37.74
Date:                Tue, 28 Sep 2021   Prob (F-statistic):           7.94e-21
Time:                        16:30:04   Log-Likelihood:                -646.93
No. Observations:                 133   AIC:                             1304.
Df Residuals:                     128   BIC:                             1318.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                                               coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------

Let's look at the results of our model. First, we can see that the overall significance test is above the threshold. In the first part of the table, you can see the F-statistic and the probability associated with it [Prob (F-statistic)]. The probability for this test is less than the conventional criterion, 0.05, so we can proceed with the interpretation of the model.

The goodness of fit test, R-squared, tells us the proportion of variation in aggregate happiness (our DV) that we can explain with the model. In this case, 68% is a good fit. The five predictors do a good job explaining variation among countries on our DV.

Next, we look at the significance tests for the individual predictors. In this case, the probabilities (show in the P>|t| column in the second part of the table) are  less than the conventional criterion, 0.05, on some independent variables but not others. We can interpret the unique effects only for the predictors that are reliable, as indicated by the significance test.

The coefficients (coef) are in the units of the dependent variable. In this case, with our DV, that would be ranks. So each coefficient tells us the change in aggregate happiness rank (which we can call "steps" or ranks) we would predict with a one unit increase in that independent variable.

The predictors that don't have a significant unique effect are still important in the model, since they serve as control variables for the significan unique effects.

We can use the results to formulate a sociological argument about the relationship between gender equality in nations and aggregate happiness.

In [ ]:
X.corr()

const  ...  Life Expectancy at Birth (Female)
const                                       NaN  ...                                NaN
Mean Years of Education (Female)            NaN  ...                           0.755096
Labour Force Participation Rate (Female)    NaN  ...                          -0.279472
Adolescent Birth Rate                       NaN  ...                          -0.737655
Percent Representation in Parliament        NaN  ...                           0.165614
Life Expectancy at Birth (Female)           NaN  ...                           1.000000

[6 rows x 6 columns]

In [ ]:
# Code block 4b: Using the merged data in a linear model
Y = GlobalIndicatorsTotal['Ladder'] # A measure of overall quality of life
X = GlobalIndicatorsTotal['HDI Rank']
X = sm.add_constant(X)
model2 = sm.OLS(Y, X, missing='drop').fit()
print(model2.summary())

                            OLS Regression Results                            
Dep. Variable:                 Ladder   R-squared:                       0.698
Model:                            OLS   Adj. R-squared:                  0.696
Method:                 Least Squares   F-statistic:                     307.6
Date:                Tue, 28 Sep 2021   Prob (F-statistic):           2.12e-36
Time:                        16:23:43   Log-Likelihood:                -628.91
No. Observations:                 135   AIC:                             1262.
Df Residuals:                     133   BIC:                             1268.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         15.0138      4.151      3.617      0.0

## More on Merging Data Sources

In this example, I'm combining a number of data files from the NYC Department of Education. They are all using the same aggregate units: schools. The DoE has a standardized identifier for schools in the system. It isn't always given the same name, but if we identify it in the data source, we can use it as the basis for zipping the files together. These files come from the NYC OpenData archive, https://opendata.cityofnewyork.us/ . In this instance, they are in JSON format.

In [ ]:
# Code block 5a: Reading a json file
DOEgraduation = pd.read_json('https://data.cityofnewyork.us/resource/nb39-jx2v.json')
DOEgraduation.head()

dbn                 school_name  ... dropped_out dropped_out_of_cohort
0  01M292  ORCHARD COLLEGIATE ACADEMY  ...         7.0                  19.4
1  01M292  ORCHARD COLLEGIATE ACADEMY  ...        10.0                  22.7
2  01M292  ORCHARD COLLEGIATE ACADEMY  ...         7.0                   9.6
3  01M292  ORCHARD COLLEGIATE ACADEMY  ...        17.0                  27.9
4  01M292  ORCHARD COLLEGIATE ACADEMY  ...         8.0                   9.4

[5 rows x 25 columns]

In [ ]:
# Code block 5b: Reading another json file
DOEregents = pd.read_json('https://data.cityofnewyork.us/resource/cbrh-qrk4.json')
DOEregents.head()

lookup  ... percent_scoring_cr
0  01M0342017Common Core Algebra  ...                 na
1   01M0342015Living Environment  ...                 na
2   01M0342016Living Environment  ...                 na
3  01M1402016Common Core Algebra  ...                 na
4  01M1402017Common Core Algebra  ...                 na

[5 rows x 17 columns]

In [ ]:
# Code block 5c: Combining the DoE json files using the merge method in pandas
DOEdata = pd.merge(left=DOEgraduation, right=DOEregents, left_on='dbn', right_on='school_dbn')
DOEdata.head()

dbn                 school_name  ... number_scoring_cr percent_scoring_cr
0  01M292  ORCHARD COLLEGIATE ACADEMY  ...                na                 na
1  01M292  ORCHARD COLLEGIATE ACADEMY  ...                na                 na
2  01M292  ORCHARD COLLEGIATE ACADEMY  ...                 0                0.0
3  01M292  ORCHARD COLLEGIATE ACADEMY  ...                15               30.0
4  01M292  ORCHARD COLLEGIATE ACADEMY  ...                29               54.7

[5 rows x 42 columns]

In [ ]:
# Code block 5d: Reading another json file
DOEdemographics = pd.read_json('https://data.cityofnewyork.us/resource/s52a-8aq6.json') # https://data.cityofnewyork.us/Education/2013-2018-Demographic-Snapshot-School/s52a-8aq6
DOEdemographics.head()

dbn                school_name  ... poverty_2  economic_need_index
0  01M015  P.S. 015 Roberto Clemente  ...      90.0              No Data
1  01M015  P.S. 015 Roberto Clemente  ...      92.3                93.5%
2  01M015  P.S. 015 Roberto Clemente  ...      84.7                89.6%
3  01M015  P.S. 015 Roberto Clemente  ...      85.4                89.2%
4  01M015  P.S. 015 Roberto Clemente  ...      84.7                89.0%

[5 rows x 39 columns]

In [ ]:
# Code block 5e: Combining the DataFrames for analysis
DOEdata2 = pd.merge(left=DOEdata, right=DOEdemographics, left_on='dbn', right_on='dbn')
DOEdata2.head()

dbn               school_name_x  ... poverty_2 economic_need_index
0  01M292  ORCHARD COLLEGIATE ACADEMY  ...      83.9             No Data
1  01M292  ORCHARD COLLEGIATE ACADEMY  ...      89.0               83.9%
2  01M292  ORCHARD COLLEGIATE ACADEMY  ...      88.1               86.7%
3  01M292  ORCHARD COLLEGIATE ACADEMY  ...      91.4               84.4%
4  01M292  ORCHARD COLLEGIATE ACADEMY  ...      91.2               89.1%

[5 rows x 80 columns]

## Data Cleaning

In [ ]:
# Code block 6a: Data cleaning
# These are the columns (variables) that are typed as objects, which in this context means strings (text). So if any of the variables we expect to
# be numeric show up here, we need to do some data cleaning.
DOEdata2.select_dtypes(object)


dbn               school_name_x  ...   year_y economic_need_index
0       01M292  ORCHARD COLLEGIATE ACADEMY  ...  2013-14             No Data
1       01M292  ORCHARD COLLEGIATE ACADEMY  ...  2014-15               83.9%
2       01M292  ORCHARD COLLEGIATE ACADEMY  ...  2015-16               86.7%
3       01M292  ORCHARD COLLEGIATE ACADEMY  ...  2016-17               84.4%
4       01M292  ORCHARD COLLEGIATE ACADEMY  ...  2017-18               89.1%
...        ...                         ...  ...      ...                 ...
139081  02M374   GRAMERCY ARTS HIGH SCHOOL  ...  2013-14             No Data
139082  02M374   GRAMERCY ARTS HIGH SCHOOL  ...  2014-15               66.6%
139083  02M374   GRAMERCY ARTS HIGH SCHOOL  ...  2015-16               64.9%
139084  02M374   GRAMERCY ARTS HIGH SCHOOL  ...  2016-17               64.1%
139085  02M374   GRAMERCY ARTS HIGH SCHOOL  ...  2017-18               75.5%

[139086 rows x 23 columns]

In [ ]:
# Code block 6b: More data cleaning
DOEdata2.loc[DOEdata2.percent_scoring_80_or_above == 's', ['percent_scoring_80_or_above']] = np.nan # Here, we're turning a string to the standard numeric missing value

DOEdata2['percent_scoring_80_or_above'] = DOEdata2['percent_scoring_80_or_above'].apply(pd.to_numeric, downcast='float', errors='coerce')
# Now we can convert the variable from a string to a numeric using the to_numeric method in pandas and applying the results to the DataFrame

DOEdata2['mean_score'] = DOEdata2['mean_score'].apply(pd.to_numeric, downcast='float', errors='coerce')
# We need to convert any string variables (objects) that we want to use in our linear model



In [ ]:
# Code block 7: Another linear model
Y = DOEdata2['mean_score']
X = DOEdata2[['poverty_2', 'english_language_learners_2', 'total_enrollment']]
X = sm.add_constant(X)
model1 = sm.OLS(Y, X, missing='drop').fit()
print(model1.summary())

                            OLS Regression Results                            
Dep. Variable:             mean_score   R-squared:                       0.382
Model:                            OLS   Adj. R-squared:                  0.382
Method:                 Least Squares   F-statistic:                 2.442e+04
Date:                Tue, 27 Oct 2020   Prob (F-statistic):               0.00
Time:                        01:02:40   Log-Likelihood:            -4.2547e+05
No. Observations:              118566   AIC:                         8.509e+05
Df Residuals:                  118562   BIC:                         8.510e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                                  coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------
const             

## Exercise

1. Read in these data files from the Department of Education, using the standardized school ID variable, as shown in the parts of code block 5.

2. Clean the data by converting the variables that should be numeric.

3. Create a linear model to explain some aspect of these education data.

4. Write a one paragraph explanation of the insights from your model.

In [15]:
import pandas as pd
import statsmodels.api as sm

# Code block 5a: Reading a json file
DOEgraduation = pd.read_json('https://data.cityofnewyork.us/resource/nb39-jx2v.json')

# Code block 5b: Reading another json file
DOEregents = pd.read_json('https://data.cityofnewyork.us/resource/cbrh-qrk4.json')

# Code block 5c: Combining the DoE json files using the merge method in pandas
DOEdata = pd.merge(left=DOEgraduation, right=DOEregents, left_on='dbn', right_on='school_dbn')

# Code block 5d: Reading another json file
DOEdemographics = pd.read_json('https://data.cityofnewyork.us/resource/s52a-8aq6.json')

# Code block 5e: Combining the DataFrames for analysis
DOEdata2 = pd.merge(left=DOEdata, right=DOEdemographics, left_on='dbn', right_on='dbn')

# Example data type conversion for 'total_enrollment' and 'mean_score' columns
DOEdata2['total_enrollment'] = pd.to_numeric(DOEdata2['total_enrollment'], errors='coerce')
DOEdata2['mean_score'] = pd.to_numeric(DOEdata2['mean_score'], errors='coerce')

# Remove rows with missing data
DOEdata2.dropna(subset=['total_grads', 'mean_score', 'total_enrollment'], inplace=True)

# Define the dependent and independent variables
y = DOEdata2['total_grads']
X = DOEdata2[['mean_score', 'total_enrollment']]

# Add a constant to the independent variables (intercept)
X = sm.add_constant(X)

# Fit the linear regression model
model = sm.OLS(y, X).fit()

# Print the model summary
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:            total_grads   R-squared:                       0.275
Model:                            OLS   Adj. R-squared:                  0.275
Method:                 Least Squares   F-statistic:                 2.103e+04
Date:                Sun, 01 Oct 2023   Prob (F-statistic):               0.00
Time:                        01:35:24   Log-Likelihood:            -5.2793e+05
No. Observations:              110725   AIC:                         1.056e+06
Df Residuals:                  110722   BIC:                         1.056e+06
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const               -6.9135      0.561  

Based on mean regents exam scores and total enrollment, the linear regression model was developed to explain the total number of graduates in NYC schools. According to the model, there is a statistically significant positive link between mean scores and graduation rate. For example, if we hold the overall enrollment constant, we can anticipate an increase of 0.0240 grads for every point with higher mean scores. Similar to the last example, there is a correlation between total enrolment and the quantity of graduates, indicating that institutions with larger enrollments typically produce more graduates. The R-squared value of 0.275 indicates that these two factors can account for 0.275 of the variation in the total number of graduates. Overall, this model can assist educational policymakers and administrators.